src/smpp_socket.erl

```diff
@@ -304,20 +300,70 @@ bound(info, {tcp_error, Sock, Reason}, #{socket := Sock} = State) ->
     reconnect({inet, Reason}, ?FUNCTION_NAME, State);
 bound(timeout, keepalive, #{role := esme} = State) ->
     State1 = send_req(State, #enquire_link{}),
-    next_state(?FUNCTION_NAME, State1);
+    keep_state(State1, [postpone]);
 bound(timeout, _, State) ->
     reconnect(timeout, ?FUNCTION_NAME, State);
 bound({call, From}, {send_req, Body, Time}, State) ->
-    State1 = send_req(State, Body, From, Time),
-    next_state(?FUNCTION_NAME, State1);
-bound(cast, {send_req, Body, Fun, Time}, State) ->
-    State1 = send_req(State, Body, Fun, Time),
-    next_state(?FUNCTION_NAME, State1);
+    bound(cast, {send_req, Body, From, Time}, State);
+bound(cast, {send_req, Body, Sender, Time}, State) ->
+    case current_time() > Time of
+        true ->
+            State1 = reply(State, {error, sending_expired}, current_time(), Sender),
+            keep_state(State1);
+        false ->
+            case check_limits(State) of
+                {ok, State1} ->
+                    State2 = send_req(State1, Body, Sender, Time),
+                    keep_state(State2);
+                {error, {NextState, State1, Actions}} ->
+                    next_state(NextState, State1, Actions)
+            end
+    end;
 bound(cast, stop, State) ->
     {stop, normal, State};
 bound(EventType, Msg, State) ->
     State1 = callback(handle_event, EventType, Msg, ?FUNCTION_NAME, State),
-    next_state(?FUNCTION_NAME, State1).
+    keep_state(State1).
+
+-spec rate_limit_cooldown(gen_statem:event_type(), term(), state()) ->
+          gen_statem:event_handler_result(statename()).
+rate_limit_cooldown(info, {tcp, Sock, Data}, #{socket := Sock} = State) ->
+    decode(Data, ?FUNCTION_NAME, State);
+rate_limit_cooldown(timeout, keepalive, #{role := esme} = State) ->
+    keep_state(State, [{next_event, timeout, keepalive}]);
+rate_limit_cooldown(state_timeout, rate_limit_reset, State) ->
+    next_state(bound, State);
+rate_limit_cooldown({call, From}, {send_req, Body, Time}, State) ->
+    rate_limit_cooldown(cast, {send_req, Body, From, Time}, State);
+rate_limit_cooldown(cast, {send_req, _Body, Fun, Time}, State) ->
+    case current_time() > Time of
+        true ->
+            State1 = reply(State, {error, rate_limit}, current_time(), Fun),
+            keep_state(State1);
+        false ->
+            keep_state(State, [postpone])
+    end;
+rate_limit_cooldown(EventType, EventContent, State) ->
+    bound(EventType, EventContent, State).
+
+-spec in_flight_limit_cooldown(gen_statem:event_type(), term(), state()) ->
+          gen_statem:event_handler_result(statename()).
+in_flight_limit_cooldown(info, {tcp, Sock, Data}, #{socket := Sock} = State) ->
+    decode(Data, ?FUNCTION_NAME, State);
+in_flight_limit_cooldown(timeout, keepalive, #{role := esme} = State) ->
+    keep_state(State, [{next_event, timeout, keepalive}]);
```

Comment:
I'd disregard _in-flight_ limit in this case actually and send the `enquire_link` request right away. I.e. keep `bound()` state function behaviour. I.e. just remove this specific function close from `in_flight_limit_cooldown`)

Reply:
Well, if we're to take the `in_flight` limit seriously (e.g. say in_flight = seq capacity), then this code is still valid. Hence,I'm dropping the comment.